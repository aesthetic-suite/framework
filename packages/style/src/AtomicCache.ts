import { CacheItem, StyleParams } from './types';

// https://jsperf.com/javascript-objects-vs-map-performance
// https://jsperf.com/performance-of-map-vs-object
export default class AtomicCache {
  cache: { [property: string]: { [value: string]: CacheItem[] } } = {};

  match(item: CacheItem, params: StyleParams, minimumRank?: number): boolean {
    if (minimumRank !== undefined && item.rank < minimumRank) {
      return false;
    }

    if (item.selector !== params.selector) {
      return false;
    }

    if (item.type !== params.type) {
      return false;
    }

    if (Array.isArray(params.conditions) && params.conditions.length > 0) {
      return params.conditions.every(i =>
        item.conditions.find(p => p.query === i.query && p.type === i.type),
      );
    }

    return true;
  }

  read(
    property: string,
    value: string,
    params: StyleParams,
    minimumRank?: number,
  ): CacheItem | null {
    const propertyCache = this.cache[property];

    if (!propertyCache) {
      return null;
    }

    const valueCache = propertyCache[value];

    if (!valueCache || valueCache.length === 0) {
      return null;
    }

    return valueCache.find(item => this.match(item, params, minimumRank)) ?? null;
  }

  write(property: string, value: string, item: CacheItem): this {
    let propertyCache = this.cache[property];

    if (!propertyCache) {
      propertyCache = {};
      this.cache[property] = propertyCache;
    }

    let valueCache = propertyCache[value];

    if (!valueCache) {
      valueCache = [];
      propertyCache[value] = valueCache;
    }

    valueCache.push(item);

    return this;
  }
}
