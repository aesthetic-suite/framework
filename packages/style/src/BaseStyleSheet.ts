import { isSSR } from '@aesthetic/utils';
import { StyleRule } from './types';

export default abstract class BaseStyleSheet {
  lastIndex: number = -1;

  sheet: StyleRule;

  protected rafHandle: number = 0;

  protected ruleBuffer: { index: number; rule: string }[] = [];

  constructor(sheet: StyleRule) {
    this.sheet = sheet;
  }

  protected enqueueRule(rule: string, customIndex?: number): number {
    let index = 0;

    // When testing or SSR, just insert the rule immediately
    if (process.env.NODE_ENV === 'test' || isSSR()) {
      index = customIndex ?? this.sheet.cssRules.length;

      this.injectRule(rule, index);

      // Otherwise buffer the insert until the next animation frame
    } else {
      index = customIndex ?? (this.lastIndex += 1);

      this.ruleBuffer.push({
        index,
        rule,
      });

      if (!this.rafHandle) {
        this.rafHandle = requestAnimationFrame(this.flushRules);
      }
    }

    return index;
  }

  protected injectRule = (rule: string, index: number) => {
    // CSS is locked in developer tools when using `insertRule`,
    // so avoid using it in development so we have full control.
    // istanbul ignore next
    if (process.env.NODE_ENV === 'development' && !isSSR()) {
      this.sheet.textContent += rule;
    } else {
      this.sheet.insertRule(rule, index);
      this.lastIndex = this.sheet.cssRules.length;
    }
  };

  protected flushRules = () => {
    this.rafHandle = 0;
    this.ruleBuffer.forEach((buffer) => {
      this.injectRule(buffer.rule, buffer.index);
    });
  };
}
