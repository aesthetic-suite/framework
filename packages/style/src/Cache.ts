import { Value, ValueWithFallbacks } from '@aesthetic/types';
import { CacheManager, CacheItem, CacheStorage, RenderOptions } from './types';

export function createCacheKey(
  property: string,
  value: Value | ValueWithFallbacks,
  { selector, conditions }: RenderOptions,
): string {
  let key = `${property}:${value}`;

  if (selector) {
    key += `:${selector}`;
  }

  if (conditions && conditions.length > 0) {
    key += `:${conditions.join(':')}`;
  }

  return key;
}

export function createCacheManager(defaultItems: CacheStorage = {}): CacheManager {
  const cache: CacheStorage = defaultItems;

  return {
    read(key, minimumRank) {
      const items = cache[key];

      if (!items) {
        return null;
      } else if (minimumRank === undefined) {
        return items[0];
      }

      return items.find((item) => item.rank! >= minimumRank) || null;
    },

    write(key: string, item: CacheItem) {
      const result = cache[key] || [];

      result.push(item);

      cache[key] = result;
    },
  };
}
