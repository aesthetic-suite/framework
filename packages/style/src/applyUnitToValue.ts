import { isUnitlessProperty } from 'css-in-js-utils';
import { Value } from './types';

export default function applyUnitToValue(property: string, value: Value): string {
  if (typeof value === 'string' || isUnitlessProperty(property) || value === 0) {
    return String(value);
  }

  return `${value}px`;
}
