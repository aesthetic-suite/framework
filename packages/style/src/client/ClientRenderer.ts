import { arrayLoop, objectLoop } from '@aesthetic/utils';
import Renderer from '../Renderer';
import GlobalStyleSheet from '../GlobalStyleSheet';
import ConditionsStyleSheet from '../ConditionsStyleSheet';
import StandardStyleSheet from '../StandardStyleSheet';
import formatVariableName from '../helpers/formatVariableName';
import getDocumentStyleSheet from '../helpers/getDocumentStyleSheet';
import { hydrateGlobals, hydrateRules, hydrateConditions } from './hydrateStyles';
import { SheetType, StyleRule, CSSVariables } from '../types';

function getSheet(type: SheetType): StyleRule {
  return (getDocumentStyleSheet(type) as unknown) as StyleRule;
}

export default class ClientRenderer extends Renderer {
  protected globalStyleSheet = new GlobalStyleSheet(getSheet('global'));

  protected conditionsStyleSheet = new ConditionsStyleSheet(getSheet('conditions'));

  protected standardStyleSheet = new StandardStyleSheet(getSheet('standard'));

  applyRootVariables(vars: CSSVariables) {
    const root = document.documentElement;

    objectLoop(vars, (value, key) => {
      root.style.setProperty(formatVariableName(key), String(value));
    });
  }

  hydrateStyles() {
    arrayLoop(
      document.querySelectorAll<HTMLStyleElement>('style[data-aesthetic-hydrate]'),
      (style) => {
        const sheet = style.sheet as CSSStyleSheet;
        const type = style.getAttribute('data-aesthetic-type') as SheetType;

        if (type === 'global') {
          hydrateGlobals(this, sheet);
        } else if (type === 'standard') {
          hydrateRules(this, sheet);
        } else if (type === 'conditions') {
          hydrateConditions(this, sheet);
        }

        // Persist the rule index
        if (!this.ruleIndex) {
          this.ruleIndex = Number(style.getAttribute('data-aesthetic-hydrate'));
        }

        // Disable so that we avoid unnecessary hydration
        style.removeAttribute('data-aesthetic-hydrate');
      },
    );
  }
}
