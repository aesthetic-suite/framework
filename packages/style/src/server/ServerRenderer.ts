import { Variables } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import Renderer from '../Renderer';
import StyleSheet from '../StyleSheet';
import ConditionsStyleSheet from '../ConditionsStyleSheet';
import TransientStyleRule from './TransientStyleRule';
import formatVariableName from '../helpers/formatVariableName';
import createStyleElement from './createStyleElement';
import { STYLE_RULE } from '../constants';

export default class ServerRenderer extends Renderer {
  conditions = new ConditionsStyleSheet('conditions', new TransientStyleRule(STYLE_RULE));

  globals = new StyleSheet('global', new TransientStyleRule(STYLE_RULE));

  standards = new StyleSheet('standard', new TransientStyleRule(STYLE_RULE));

  applyRootVariables(vars: Variables) {
    objectLoop(vars, (value, key) => {
      this.globals.sheet.cssVariables[formatVariableName(key)] = String(value);
    });
  }

  extractStyles<T>(result: T): T {
    global.AESTHETIC_CUSTOM_RENDERER = this;

    return result;
  }

  renderToStyleMarkup(): string {
    return (
      createStyleElement('global', this.getRootRule('global'), this.ruleIndex) +
      createStyleElement('standard', this.getRootRule('standard'), this.ruleIndex) +
      createStyleElement('conditions', this.getRootRule('conditions'), this.ruleIndex)
    );
  }
}
