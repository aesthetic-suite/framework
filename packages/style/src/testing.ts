/* eslint-disable unicorn/import-index */

import { CSS } from '@aesthetic/types';
import {
  CacheManager,
  StyleEngine,
  EngineOptions,
  SheetManager,
  SheetType,
  StyleRule,
  getDocumentStyleSheet,
} from './index';
import { createCacheManager, createStyleElements, createEngine, createSheetManager } from './next';

export function createTestCacheManager(): CacheManager {
  return createCacheManager();
}

export function createTestSheetManager(): SheetManager {
  return createSheetManager(createStyleElements());
}

export function createTestStyleEngine(options: Partial<EngineOptions>): StyleEngine {
  return createEngine({
    cacheManager: createTestCacheManager(),
    sheetManager: createTestSheetManager(),
    ...options,
  });
}

export function getRenderedStyles(type: SheetType | StyleRule): CSS {
  return Array.from(
    ((typeof type === 'string' ? getDocumentStyleSheet(type) : type) as StyleRule).cssRules,
  ).reduce((css, rule) => css + rule.cssText, '');
}

export function purgeStyles(type?: SheetType): void {
  if (type) {
    // This is the only way to generate accurate snapshots.
    // It may slow down tests though?
    document.getElementById(`aesthetic-${type}`)?.remove();
  } else {
    purgeStyles('global');
    purgeStyles('standard');
    purgeStyles('conditions');
  }
}
