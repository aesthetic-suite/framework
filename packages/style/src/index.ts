/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Direction, EngineOptions, VariablesMap } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import { createStyleElements, getStyleElement } from './client/dom';
import { hydrateStyles } from './client/hydration';
import { createCacheManager } from './common/cache';
import { createStyleEngine } from './common/engine';
import { createSheetManager } from './common/sheet';
import { formatVariable } from './common/syntax';
import { StyleEngine } from './types';

export * from './common/cache';
export * from './common/constants';
export * from './common/engine';
export * from './common/helpers';
export * from './common/sheet';
export * from './common/syntax';
export * from './types';

export { createStyleElements, getStyleElement };

// Set active direction on the document `dir` attribute
function setDirection(direction: Direction) {
  document.documentElement.setAttribute('dir', direction);
}

// Set CSS variables to :root
function setRootVariables(vars: VariablesMap) {
  objectLoop(vars, (value, key) => {
    document.documentElement.style.setProperty(formatVariable(key), String(value));
  });
}

// Set active theme class names on the `body`
function setTheme(classNames: string[]) {
  document.body.className = classNames.join(' ');
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  if (global.AESTHETIC_CUSTOM_ENGINE) {
    return global.AESTHETIC_CUSTOM_ENGINE as StyleEngine;
  }

  const direction = (document.documentElement.getAttribute('dir') ||
    document.body.getAttribute('dir') ||
    'ltr') as Direction;

  const engine: StyleEngine = createStyleEngine({
    cacheManager: createCacheManager(),
    direction,
    sheetManager: createSheetManager(createStyleElements()),
    ...options,
  });

  // Match against browser preferences
  engine.prefersColorScheme = (scheme) => matchMedia(`(prefers-color-scheme: ${scheme})`).matches;
  engine.prefersContrastLevel = (level) => matchMedia(`(prefers-contrast: ${level})`).matches;

  // Handle DOM specific logic
  engine.setDirection = setDirection;
  engine.setRootVariables = setRootVariables;
  engine.setTheme = setTheme;

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
