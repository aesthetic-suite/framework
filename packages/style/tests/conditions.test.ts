import ClientRenderer from '../src/client/ClientRenderer';
import ServerRenderer from '../src/server/ServerRenderer';
import { getRenderedStyles, purgeStyles } from '../src/testing';

describe('Conditions', () => {
  let renderer: ClientRenderer;

  beforeEach(() => {
    renderer = new ClientRenderer();
  });

  afterEach(() => {
    purgeStyles();
  });

  describe('media queries', () => {
    it('supports @media conditions', () => {
      const className = renderer.renderRule({
        background: '#000',
        padding: '15px',
        '@media (max-width: 600px)': {
          padding: '15px',
        },
        '@media screen and (min-width: 900px)': {
          padding: '20px',
        },
      });

      expect(className).toBe('a b c d');
      expect(getRenderedStyles('standard')).toMatchSnapshot();
      expect(getRenderedStyles('conditions')).toMatchSnapshot();
    });

    it('can be nested in @supports', () => {
      const className = renderer.renderRule({
        padding: '15px',
        '@supports (display: flex)': {
          '@media (max-width: 600px)': {
            padding: '15px',
          },
        },
      });

      expect(className).toBe('a b');
      expect(getRenderedStyles('standard')).toMatchSnapshot();
      expect(getRenderedStyles('conditions')).toMatchSnapshot();
    });

    it('sorts media queries using mobile-first (when using SSR)', () => {
      const block = { padding: 0 };
      const serverRenderer = new ServerRenderer();

      serverRenderer.renderRule({
        '@media screen and (min-width: 1024px)': block,
        '@media screen and (min-width: 320px) and (max-width: 767px)': block,
        '@media screen and (min-width: 1280px)': block,
        '@media screen and (min-height: 480px)': block,
        '@media screen and (min-height: 480px) and (min-width: 320px)': block,
        '@media screen and (orientation: portrait)': block,
        '@media screen and (min-width: 640px)': block,
        '@media print': block,
        '@media screen and (max-width: 767px) and (min-width: 320px)': block,
        '@media tv': block,
        '@media screen and (max-height: 767px) and (min-height: 320px)': block,
        '@media screen and (orientation: landscape)': block,
        '@media screen and (min-device-width: 320px) and (max-device-width: 767px)': block,
        '@media screen and (max-width: 639px)': block,
        '@media screen and (max-width: 1023px)': block,
      });

      expect(
        getRenderedStyles(serverRenderer.sheetManager.getSheet('conditions')),
      ).toMatchSnapshot();
    });
  });

  describe('support queries', () => {
    it('supports @supports conditions', () => {
      const className = renderer.renderRule({
        display: 'block',
        '@supports (display: flex)': {
          display: 'flex',
        },
      });

      expect(className).toBe('a b');
      expect(getRenderedStyles('standard')).toMatchSnapshot();
      expect(getRenderedStyles('conditions')).toMatchSnapshot();
    });

    it('can be nested in @media', () => {
      const className = renderer.renderRule({
        display: 'block',
        '@media screen and (min-width: 900px)': {
          '@supports (display: flex)': {
            display: 'flex',
          },
        },
      });

      expect(className).toBe('a b');
      expect(getRenderedStyles('standard')).toMatchSnapshot();
      expect(getRenderedStyles('conditions')).toMatchSnapshot();
    });
  });
});
