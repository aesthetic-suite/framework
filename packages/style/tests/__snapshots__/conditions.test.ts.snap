// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conditions media queries can be nested in @supports 1`] = `".a {padding: 15px;}"`;

exports[`Conditions media queries can be nested in @supports 2`] = `"@supports (display: flex) {@media (max-width: 600px) {.b {padding: 15px;}}}"`;

exports[`Conditions media queries sorts media queries using mobile-first (when using SSR) 1`] = `""`;

exports[`Conditions media queries supports @media conditions 1`] = `".a {background: #000;}.b {padding: 15px;}"`;

exports[`Conditions media queries supports @media conditions 2`] = `"@media (max-width: 600px) {.c {padding: 15px;}}@media screen and (min-width: 900px) {.d {padding: 20px;}}"`;

exports[`Conditions support queries can be nested in @media 1`] = `".a {display: block;}"`;

exports[`Conditions support queries can be nested in @media 2`] = `"@media screen and (min-width: 900px) {@supports (display: flex) {.b {display: flex;}}}"`;

exports[`Conditions support queries supports @supports conditions 1`] = `".a {display: block;}"`;

exports[`Conditions support queries supports @supports conditions 2`] = `"@supports (display: flex) {.b {display: flex;}}"`;
